### DISCUSSION TITLE
Hasura connection error is not reported correctly

### DISCUSSION THREAD

**Discussion by lovasoa** [enisdenjo/graphql-ws#500] [Q&A] [answered]
**Screenshot**
![image](https://github.com/enisdenjo/graphql-ws/assets/552629/0ca82892-d44f-4828-b5db-17fccbefa87c)

### Reproduction

Connecting to a hasura graphql server with a jwt that does not contain the `x-hasura-default-role` claim.

**Expected Behaviour**
The error returned by hasura is properly forwarded to the user. 

Hasura returns `{"type":"connection_error","payload":"JWT claim does not contain x-hasura-default-role"}`

The error message returned by the library should contain the string `JWT claim does not contain x-hasura-default-role`

**Actual Behaviour**
but instead the library threw the following error : 

```
 Error: Invalid message 'type' property "connection_error"
    validateMessage common.mjs:142
    parseMessage common.mjs:168
    onmessage client.mjs:214
    connecting client.mjs:212
    connecting client.mjs:256
```

**Comment by enisdenjo** [comment, marked as answer]
There is no `connection_error` message type in the [GraphQL over WebSocket spec](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md#message-types).

This is a problem with Hasura, please open an issue there.

**Reply by lovasoa** [reply]
Even if hasura changes that in the future, there are tens of thousands of deployed instances in the wild with that behavior. Hasura is quite popular. Are you sure you don't want to add support for this `connection_error` key, even if it's not in the standard ? 

And if not, maybe we could at least change the error message to something that would include the original message. That would already help a lot. Maybe : `The following message was received: {"type":"connection_error","payload":"JWT claim does not contain x-hasura-default-role"}, but it could not be decoded because the type "connection_error" is invalid.`

**Reply by lovasoa** [reply]
Looking at the link you sent, it looks like the specification is your own, isn't it ? Is there a standards body involved ? If not, isn't the easiest solution to just add `connection_error` to your own  `PROTOCOL.md` ?

**Reply by enisdenjo** [reply]
> Are you sure you don't want to add support for this connection_error key, even if it's not in the standard ?

Yes. Standards should not adapt to implementations, it should be the other way around.

Furthermore, there's a good reason for not having the `connection_error` - I explained a while ago on the same issue you referenced in Hasura: https://github.com/enisdenjo/graphql-ws/issues/112#issuecomment-779950932.

> [...] there are tens of thousands of deployed instances in the wild with that behavior.

There's also tens of thousands instances of other services and projects that don't use the the `connection_error` and comply with the spec.

> Looking at the link you sent, it looks like the specification is your own, isn't it ? Is there a standards body involved ?

Yes. The [official GraphQL over HTTP work group](https://github.com/graphql/graphql-over-http). Also see the [RFC there](https://github.com/graphql/graphql-over-http/blob/main/rfcs/GraphQLOverWebSocket.md).
